### PR TITLE
LockstakeClipper Callee

### PR DESCRIPTION
**Pull request by valiafetisov**
This PR implements callee contract for the [LockstakeClipper](https://github.com/makerdao/lockstake/blob/dev/src/LockstakeClipper.sol) using UniswapV2. It allows anyone to participate in auctions using UniswapV2 path to exchange MKR (or NGT) into DAI (or NST).

Notes:

- CI is expected to fail due to the tests for other callees being fragile / depending on the market
  - To test this specific callee, one can run `forge test --match-contract=UniswapV2LockstakeCalleeTest` locally
- In order to write tests for the new callee, we had to update dependencies and remove hardcoded solidity version at the foundry level (`--use solc:0.6.12`) while strictly setting it in every callee contract instead. This is due to [LockstakeEngine requiring at least 0.8.21](https://github.com/makerdao/lockstake/blob/39dbea91e47911ddbf1122d5edd8c4a99934b059/src/LockstakeEngine.sol#L17) solidity version
  - If requested, those changes could be moved into a separate maintenance PR to keep git history clean
- Current version of the callee support any UniV2 paths that start from MKR or NGT and end with DAI or NST (in other words both MKR/DAI and NGT/NST pools are supported, but not only). The previous version of the callee (found at 9b5cb11) required a redeployment to support different destination token
  - Please let us know which contract design is preferable

### CODE REVIEW

File: src/UniswapV2LockstakeCallee.sol [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.21;
+
+interface TokenLike {
+    function approve(address, uint256) external;
+    function transfer(address, uint256) external;
+    function balanceOf(address) external view returns (uint256);
+    function decimals() external view returns (uint256);
```
Comment (oldchili) [inline src/UniswapV2LockstakeCallee.sol:24]: Unused

File: src/UniswapV2LockstakeCallee.sol [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+interface UniswapV2Router02Like {
+    function swapExactTokensForTokens(uint256, uint256, address[] calldata, address, uint256) external returns (uint[] memory);
```
Comment (oldchili) [inline src/UniswapV2LockstakeCallee.sol:28]: uint => uint256

File: src/UniswapV2LockstakeCallee.sol [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+interface NstJoinLike {
+    function nst() external view returns (address);
+    function join(address, uint256) external;
+}
+
+interface MkrNgt {
+    function mkr() external view returns (address);
+    function ngt() external view returns (address);
+    function rate() external view returns (uint256);
+    function mkrToNgt(address usr, uint256 mkrAmt) external;
```
Comment (oldchili) [inline src/UniswapV2LockstakeCallee.sol:45]: Parameters here are named as opposed to other places.

File: src/UniswapV2LockstakeCallee.sol [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        ngt = TokenLike(mkrNgt.ngt());
+    }
+
+    function divup(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x != 0 ? ((x - 1) / y) + 1 : 0;
+    }
+
+    function clipperCall(
+        address sender,     // Clipper Caller and DAI/NST delivery address
+        uint256 dstAmt,     // DAI/NST amount to payback[rad]
```
Comment (oldchili) [inline src/UniswapV2LockstakeCallee.sol:81]: missing space before [rad]

File: src/UniswapV2LockstakeCallee.sol [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        uint256 gemAmt,     // Gem amount received [wad]
+        bytes calldata data // Extra data needed (gemJoin)
```
Comment (oldchili) [inline src/UniswapV2LockstakeCallee.sol:83]: Suggest to remove "(gemJoin)"

**Review by oldchili — COMMENTED** [review]
Only looked at the new callee contract, looks good.
Made some minor comments.

**Comment by oldchili** [comment]
All changes look good, I'm not able to resolve the comments, so please do that on your side.

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+    MockUniswapRouter02      uniRouter02;
+    uint256                  uniV2Price = 1;
+    address[]                uniV2Path = new address[](2);
+    UniswapV2LockstakeCallee callee;
+
+    event OnKick(address indexed urn, uint256 wad);
+    event OnTake(address indexed urn, address indexed who, uint256 wad);
+    event OnRemove(address indexed urn, uint256 sold, uint256 burn, uint256 refund);
+
+    // Match https://github.com/makerdao/lockstake/blob/735e1e85ca706534a77d8e1582df0d3248cbd2b6/test/LockstakeEngine.t.sol#L87-L177
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:94]: nit, 
it doesn't fully match with the commented link as `dai` is missing there. 

https://www.diffchecker.com/5eGEigBq/

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+            id,                // Auction id
+            20_000 * 10**18,   // Upper limit on amount of collateral to buy  [wad]
+            type(uint256).max, // Maximum acceptable price (DAI / collateral) [ray]
+            address(callee),   // Receiver of collateral and external call address
+            flashData          // Data to pass in external call; if length 0, no call is done
+        );
+        assertEq(mkr.balanceOf(address(callee)), 0, "invalid-callee-mkr-balance");
+        assertEq(dst.balanceOf(address(callee)), 0, "invalid-callee-dst-balance");
+        assertEq(mkr.balanceOf(buyer1), 0, "invalid-final-buyer2-mkr-balance");
+        assertEq(dst.balanceOf(buyer1), expectedProfit, "invalid-final-buyer2-dst-balance");
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:300]: the error messages need to be updated to `buyer1`

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+
+        // Setup different buyer to take the rest of the auction
+        address buyer2 = address(222);
+        vm.prank(buyer2); dss.vat.hope(address(clip));
+        assertEq(mkr.balanceOf(buyer2), 0, "unexpected-initial-buyer1-mkr-balance");
+        assertEq(dst.balanceOf(buyer2), 0, "unexpected-initial-buyer1-dst-balance");
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:306]: the error messages need to be updated to `buyer2`

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+        uint buyAmt = amountIn * fixedPrice;
+        require(amountOutMin <= buyAmt, "Minimum Fill not reached");
+
+        uint256 initialInBalance = DSTokenAbstract(path[0]).balanceOf(address(this));
+        DSTokenAbstract(path[0]).transferFrom(msg.sender, address(this), amountIn);
+        assertEq(DSTokenAbstract(path[0]).balanceOf(address(this)), initialInBalance + amountIn);
+
+        uint256 initialOutBalance = DSTokenAbstract(path[path.length - 1]).balanceOf(msg.sender);
+        GodMode.setBalance(path[path.length - 1], msg.sender, initialOutBalance + buyAmt);
+        assertEq(DSTokenAbstract(path[path.length - 1]).balanceOf(msg.sender), initialOutBalance + buyAmt);
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:46]: nit, would it make sense to send `outBalance` to `to` address?

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+        TokenLike dst = TokenLike(path[path.length - 1]);
+
+        // Setup buyer
+        address buyer1 = address(111);
+        vm.prank(buyer1); dss.vat.hope(address(clip));
+        assertEq(mkr.balanceOf(buyer1), 0, 'unexpected-initial-buyer1-mkr-balance');
+        assertEq(dst.balanceOf(buyer1), 0, 'unexpected-initial-buyer1-dst-balance');
+
+        // Partial profit
+        uint256 expectedProfit = (20_000 * 10**18) * exchangeRate - 20_000 * pip.read() * clip.buf() / RAY;
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:267]: nit,
For me, it was a bit confusing where this `20_000` came from without reading further.
I think it'd be nice if we could put this value in a var (eg. `amountToBy1 = 20_000`)

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+        assertEq(dst.balanceOf(address(callee)), 0, "invalid-callee-dst-balance");
+        assertEq(mkr.balanceOf(buyer1), 0, "invalid-final-buyer2-mkr-balance");
+        assertEq(dst.balanceOf(buyer1), expectedProfit, "invalid-final-buyer2-dst-balance");
+
+        // Setup different buyer to take the rest of the auction
+        address buyer2 = address(222);
+        vm.prank(buyer2); dss.vat.hope(address(clip));
+        assertEq(mkr.balanceOf(buyer2), 0, "unexpected-initial-buyer1-mkr-balance");
+        assertEq(dst.balanceOf(buyer2), 0, "unexpected-initial-buyer1-dst-balance");
+        address profitAddress = address(333);
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:307]: Can we also add this check for profitAddress to ensure the sanity of L326? 
```sol
 assertEq(dst.balanceOf(profitAddress), 0, "unexpected-initial-profit-address-dst-balance");
```

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+        );
+        assertEq(mkr.balanceOf(address(callee)), 0, "invalid-callee-mkr-balance");
+        assertEq(dst.balanceOf(address(callee)), 0, "invalid-callee-dst-balance");
+        assertEq(mkr.balanceOf(buyer2), 0, "invalid-final-buyer2-mkr-balance");
+        assertEq(dst.balanceOf(profitAddress), expectedProfit, "invalid-final-profit");
+    }
+
+    // --- Callee tests using MKR/DAI path ---
+
+    function testCalleeTakeNoWithStakingNoDelegateMkrDai() public {
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:331]: nit,
- Could we use `with` / `without` for unified function names?
eg. `testCalleeTakeWithoutStakingWithoutDelegateMkrDai` /`testCalleeTakeWithStakingWithDelegateMkrDai`
- For me, it'd be easier to read the test if the test function name sequence matches with `_testCalleeTake` params sequence.
eg. `testCalleeTakeWithoutDelegateWithoutStakingMkrDai`

File: src/test/UniswapV2LockstakeCallee.t.sol [hunk 13]
```diff
@@ -0,0 +1,418 @@
...
+    }
+
+    function testCalleeTakeWithStakingWithDelegateMkrNst() public {
+        setUpCallee();
+        uniV2Path[0] = address(mkr);
+        uniV2Path[1] = address(nst);
+        _testCalleeTake(true, true, uniV2Path, uniV2Price);
+    }
+
+    // --- Callee tests using NGT/NST path ---
```
Comment (DaeunYoon) [inline src/test/UniswapV2LockstakeCallee.t.sol:389]: The function titles in this section don't match with the section title & function content. 
Could you please address those?
eg. `testCalleeTakeNoWithStakingNoDelegateNgtDai` should be `testCalleeTakeNoWithStakingNoDelegateNgtNst`

**Review by DaeunYoon — APPROVED** [review]
All comments are addressed and the changes look good to me.
Feel free to resolve the above comments as I'm not able to do it.

**Review by SidestreamIcedMango — APPROVED** [review]
- [x] Checked callee contract
	- [x] Static Interfaces
	    * [x] No unused static interfaces
	    * [x] Interface naming style should match with `Like` suffix (e.g. `VatLike`)
	      * EXCEPTION: [known interface naming exceptions](https://github.com/makerdao/dss-exec-lib/blob/master/src/DssExecLib.sol#L24-L52)
	    * [x] Each static interface declare only functions actually used in the spell code
	* [x] Precision units
		* [x] Precision units match their defined values:
	      * `WAD = 10 ** 18`
	      * `RAY = 10 ** 27`
	      * `RAD = 10 ** 45`
	    * [x] Precision units match with [Numerical Ranges](https://github.com/makerdao/dss/wiki/Numerical-Ranges#notation)
	* [x] Ensure every variable is declared as `public`/`internal`
- [x] Checked tests
	- [x] Ensure solc version is matching with contract
	- [x] Checked `MockUniswapRouter02`
		- [x] Simulates `UniswapV2Router02` deployed at [0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D](https://etherscan.io/address/0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D#code)
			- address found in the [uniswap docs](https://docs.uniswap.org/contracts/v2/reference/smart-contracts/v2-deployments)
		- [x] Interface for function `swapExactTokensForTokens` matches
		- [x] Behaviour of the function is correctly mocked
	- [x] Checked `UniswapV2LockstakeCalleeTest`
		- [x] Function `setUp` matches with provided existing test function
		- [x] Function `_urnSetUp` matches with provided existing test function
		- [x] Function `_forceLiquidation` matches with provided existing test function
		- [x] Checked `setUpCallee`
			- [x] Correctly sets up test callee
		- [x] Checked `_testCalleeTake`
			- [x] Simulates and tests multiple actors using the callee to take from an auction
	- [x] Test calls cover all relevant cases
- [x] Checked formatting
- [x] Check all tests are passing locally using `make test`

```bash
Ran 16 tests for src/test/UniswapV2LockstakeCallee.t.sol:UniswapV2LockstakeCalleeTest
[PASS] testCalleeTake_NoDelegate_NoStaking_MkrDai() (gas: 3476105)
[PASS] testCalleeTake_NoDelegate_NoStaking_MkrNst() (gas: 3401278)
[PASS] testCalleeTake_NoDelegate_NoStaking_NgtDai() (gas: 3576068)
[PASS] testCalleeTake_NoDelegate_NoStaking_NgtNst() (gas: 3501285)
[PASS] testCalleeTake_NoDelegate_WithStaking_MkrDai() (gas: 3515101)
[PASS] testCalleeTake_NoDelegate_WithStaking_MkrNst() (gas: 3440273)
[PASS] testCalleeTake_NoDelegate_WithStaking_NgtDai() (gas: 3615082)
[PASS] testCalleeTake_NoDelegate_WithStaking_NgtNst() (gas: 3540253)
[PASS] testCalleeTake_WithDelegate_NoStaking_MkrDai() (gas: 3513726)
[PASS] testCalleeTake_WithDelegate_NoStaking_MkrNst() (gas: 3438942)
[PASS] testCalleeTake_WithDelegate_NoStaking_NgtDai() (gas: 3613688)
[PASS] testCalleeTake_WithDelegate_NoStaking_NgtNst() (gas: 3538862)
[PASS] testCalleeTake_WithDelegate_WithStaking_MkrDai() (gas: 3552697)
[PASS] testCalleeTake_WithDelegate_WithStaking_MkrNst() (gas: 3477869)
[PASS] testCalleeTake_WithDelegate_WithStaking_NgtDai() (gas: 3654092)
[PASS] testCalleeTake_WithDelegate_WithStaking_NgtNst() (gas: 3612389)
Suite result: ok. 16 passed; 0 failed; 0 skipped; finished in 50.59s (52.86s CPU time)
```

**Pull request merged**